### ISSUE TITLE
Correctly model OSB-specific fields in the API

### ISSUE THREAD

**Issue by pmorie** [kubernetes-incubator/service-catalog#218]
Currently, after #213, we have a number of fields in the API prefixed with `OSB` to indicate that they are for the Open Service Broker API.  Prefixing is a terrible way to handle this, and for the record, was never my plan -- just a temporary measure to isolate what is intrinsic to the kubernetes service catalog and what is a function of the OSB API.  The correct way to do this is to encapsulate api-specific fields in a subobject.  An example of an area that we already do this in is the volume API in kubernetes.  There is an API type called [`VolumeSource`](https://github.com/kubernetes/kubernetes/blob/2d803afc980ad33d0c17b68af78bd4903200b96b/pkg/api/types.go#L209) that is basically a union type which can be exactly one of a number of different volume-specific objects.  There is a validation that enforces that one and only one of these pointer fields can be set.  This gives us a way to isolate volume-type-specific fields and encapsulate them correctly.

For an example, let's look at the the `ServiceClass` object and see how we would apply that pattern.  Currently, as of this writing, that object looks like so in the API:

```go
// ServiceClass represents an offering in the service catalog.
type ServiceClass struct {
	metav1.TypeMeta
	kapi.ObjectMeta

	// BrokerName is the reference to the Broker that provides this service.
	// Immutable.
	BrokerName string

	Bindable      bool
	Plans         []ServicePlan
	PlanUpdatable bool // Do we support this?

	// OSB-specific
	// OSBGUID is the identity of this object for use with the OSB API.
	// Immutable.
	OSBGUID string

	// OSB-specific
	OSBTags                    []string
	OSBRequires                []string
	OSBMaxDBPerNode            string
	OSBMetadata                interface{}
	OSBDashboardOAuth2ClientID string
	OSBDashboardSecret         string
	OSBDashboardRedirectURI    string
}
```

Note, for the record, I'm thinking now that some of the fields in this object which aren't currently prefixed are really also specific to the API.  Know that, and let's look at what this would look like applying the pattern I described:

```go
type ServiceClass struct {
	metav1.TypeMeta
	kapi.ObjectMeta

	// BrokerName is the reference to the Broker that provides this service.
	// Immutable.
	BrokerName string

	Bindable      bool
	Plans         []ServicePlan
	PlanUpdatable bool // Do we support this?

	// Note for the record I think these names are terrible;
	// they are merely for discussion and not meant as suggestions

	APIOptions ServiceClassAPIOptions
}

// In the future, we can support other broker APIs by adding
// new pointer fields to this struct
type ServiceClassAPIOptions struct {
    OpenServiceBroker *OpenServiceBrokerAPIServiceClassOptions	
}

type OpenServiceBrokerAPIServiceClassOptions struct {
	GUID                    string
	Tags                    []string
	Requires                []string
	MaxDBPerNode            string
	Metadata                interface{}
	DashboardOAuth2ClientID string
	DashboardSecret         string
	DashboardRedirectURI    string
}
```

It's arguable that this is just an example of the single table inheritance pattern.  There are a bunch of different names for this pattern and the different pieces of it.  Anyway....

I'm posting this for discussion.  There are some things that need to be thought through like exactly which fields are really OSB-specific and which are more universal, how to deal with mismatched options in the future (IE, service class is provided by an OSB broker and someone makes an instance with $OTHER_THING options).

**Comment by krancour** [comment]
I feel like we should get #210 in before this, or else face a difficult rebase.

**Comment by pmorie** [comment]
@krancour completely agree.

**Comment by pmorie** [comment]
I am now of the opinion that this was a premature idea, and that we should not worry about this before MVP 2 is finished.  When we decide to make a `v1beta1` API, then it will be appropriate to think about this.

**Comment by pmorie** [comment]
Closed by #834
